mdal/frmts/mdal_dynamic_driver.cpp

```diff
@@ -0,0 +1,184 @@
+#include "mdal_dynamic_driver.hpp"
+#include "mdal_logger.hpp"
+
+#include <dlfcn.h>
+#include <string.h>
+
+
+static void *getSymBol( void *library, const std::string &symbolName )
+{
+  void *symbol =  dlsym( library, symbolName.c_str() );
+
+#if (defined(__APPLE__) && defined(__MACH__))
+  /* On mach-o systems, C symbols have a leading underscore and depending
+         * on how dlcompat is configured it may or may not add the leading
+         * underscore.  If dlsym() fails, add an underscore and try again.
+         */
+  if ( symbol == nullptr )
+  {
+    char withUnder[256] = {};
+    snprintf( withUnder, sizeof( withUnder ), "_%s", symbolName );
+    symbol = dlsym( library, withUnder );
+  }
+#endif
+
+  return symbol;
+}
+
+MDAL::DriverDynamic::~DriverDynamic()
+{
+  if ( mLibrary )
+    dlclose( mLibrary );
+}
+
+MDAL::Driver *MDAL::DriverDynamic::create()
+{
+  return new DriverDynamic( name(), longName(), filters(), mCapabilityFlags, mLibFilepath );
+}
+
+bool MDAL::DriverDynamic::canReadMesh( const std::string &uri )
+{
+  return mCanReadMeshFunction( uri.c_str() );
+}
+
+std::unique_ptr<MDAL::Mesh> MDAL::DriverDynamic::load( const std::string &uri, const std::string &meshName )
+{
+  MDAL_UNUSED( meshName );
+  int meshId = mOpenMeshFunction( uri.c_str() );
+  if ( meshId != -1 )
+  {
+    if ( mMeshIds.find( meshId ) != mMeshIds.end() )
+      MDAL::Log::error( MDAL_Status::Err_InvalidData, longName(), "unable to load the mesh, duplicate mesh id" );
+
+    std::unique_ptr<MDAL::MeshDynamicDriver> mesh( new MeshDynamicDriver( name(), 100, uri, meshId ) );
+    if ( mesh->loadSymbol( mLibrary ) )
+    {
+      mMeshIds.insert( meshId );
+      return mesh;
+    }
+  }
+
+  return std::unique_ptr<MDAL::Mesh>();
+}
+
+MDAL::Driver *MDAL::DriverDynamic::create( const std::string &libFile )
+{
+  void *library = dlopen( libFile.c_str(), RTLD_LAZY );
```

Comment:
create custom simple smart pointer for library pointer for dlopen and dlclose of library so we never leak the memory